### PR TITLE
Optionals - Add No Medical and No Realistic Names components

### PR DESCRIPTION
**Pull request by Dystopian**
In some missions ACE Medical can't be used. Also there were some players who wanted to disable Realistic Names component.
I think this PR is most correct way to disable ACE components without mod folder modifying and messing with own mods based on ACE. Empty config is correctly loaded and works as expected when such optional component is enabled.

The biggest problem is that build tool must support `__has_include` preprocessor command. As of now Mikero tools (DePbo 7.97.0.25) don't support it and even with `-B` argument they try to rapify/check syntax (fail as result). AddonBuilder builds successfully.

About `$NOBIN$`. It looks like Mikero tools with `-B (Don't binarise)` argument do not binarize `config.cpp` but binarize models etc. AddonBuilder with `-packonly` argument do not binarize anything. So both tools can't be used for this PR building :disappointed: 
I would suggest addding new file flag (e.g. `$NOBINCONFIG$`) to indicate that only config should not be binarized. And naturally build tool (HEMTT?) should support such mode. Config syntax should be checked anyway ignoring `__has_include`.
@BrettMayson could you implement such feature if ACE team agrees to approve this PR?

Making this PR as draft until there is a tool for building.

### CODE REVIEW

File: addons/medical/config.cpp [hunk 1]
```diff
@@ -1,3 +1,7 @@
+#if __has_include("\z\ace\addons\nomedical\script_component.hpp")
```
Comment (commy2) [inline addons/medical/config.cpp:1]: Does this even work when ACE is binned?

Comment (Dystopian) [reply addons/medical/config.cpp:1]: no, medical components must be unbinned

Comment (commy2) [reply addons/medical/config.cpp:1]: That's pretty bad.

Comment (PabstMirror) [reply addons/medical/config.cpp:1]: we're talking about configs, not models 
the difference between loading a binned vs non-binned config is probably a milisecond

Comment (jokoho48) [reply addons/medical/config.cpp:1]: The bigger issue is that hemtt does not support this yet. so using this would break the build tools

**Comment by BaerMitUmlaut** [comment]
What's the point of this? What is the advantage of having to add a file instead of having to remove it? This doesn't make sense to me, sorry.

**Comment by commy2** [comment]
Probably something about Steam. I wouldn't know.

**Comment by Dystopian** [comment]
> 
> 
> What's the point of this? What is the advantage of having to add a file instead of having to remove it? This doesn't make sense to me, sorry.

- you don't have to manage files in ACE directory on each ACE update;
- you can simply disable/enable ACE medical for some game sessions/missions with just loading/unloading optional addon.

**Comment by BaerMitUmlaut** [comment]
> you don't have to manage files in ACE directory on each ACE update;

Does the optional PBO move itself?

**Comment by commy2** [comment]
No, but you can un/check it in Steam I suppose.

**Comment by Dystopian** [comment]
> Does the optional PBO move itself?

You don't have to move optionals since #5038.

**Comment by PabstMirror** [comment]
Problem we've found is that `__has_include` only checks if the file can be loaded
but it applies to both the virtual filesystem (files from pbos) and real files in the arma folder
e.g.
`#if __has_include("\test\t.hpp")`
will return true if you have a `"F:\SteamLibrary\steamapps\common\Arma 3\test\t.hpp"`

it's just like #inclding a `\userconfig\` file
we really need a `__has_include` that only checks virtual filesystem

**Comment by PabstMirror** [comment]
preprocessor should respect filepatching in 2.06

**Comment by Superxpdude** [comment]
Are there any plans to continue development of this PR? I am interested in this as a feature, and can put in the time to get it working via another PR, but I don't want to duplicate the work in case there are still plans to finish this one up.

Mostly wondering because the limiting factor for this seems to be resolved as far as I can tell based on PR #9116.

**Comment by Dystopian** [comment]
@Superxpdude I will try to update it this week.

**Comment by jonpas** [comment]
What's still preventing Medical from having a setting to enable/disable it?

**Comment by Dystopian** [comment]
> What's still preventing Medical from having a setting to enable/disable it?

IIRC there were changes in unit damage config. I didn't find better way to handle it.

**Comment by LinkIsGrim** [comment]
> > What's still preventing Medical from having a setting to enable/disable it?
> 
> IIRC there were changes in unit damage config. I didn't find better way to handle it.

We could handle the limb armor via script in `fnc_handleDamage` (ugh), I guess. Will just adding ACE_HDBracket change vanilla behavior in anyway?

**Review by LinkIsGrim — APPROVED** [review]

File: addons/fire/addon.toml [hunk 2]
```diff
@@ -0,0 +1 @@
+preprocess = false
```
Comment (PabstMirror) [inline addons/fire/addon.toml:1]: need to update syntax
I can do this, just making note

File: optionals/nomedical/$PBOPREFIX$ [hunk 3]
```diff
@@ -0,0 +1 @@
+z\ace\addons\nomedical
```
Comment (jonpas) [inline optionals/nomedical/$PBOPREFIX$:1]:
```suggestion
z\ace\addons\nomedical

```

File: optionals/norealisticnames/$PBOPREFIX$ [hunk 4]
```diff
@@ -0,0 +1 @@
+z\ace\addons\norealisticnames
```
Comment (jonpas) [inline optionals/norealisticnames/$PBOPREFIX$:1]:
```suggestion
z\ace\addons\norealisticnames

```

**Review by jonpas — CHANGES_REQUESTED** [review]

File: optionals/norealisticnames/$PBOPREFIX$ (hunk 4 above)
Comment (LinkIsGrim) [reply optionals/norealisticnames/$PBOPREFIX$:1]: We don't have newlines at EOF for any other component. Sure about this?

Comment (jonpas) [reply optionals/norealisticnames/$PBOPREFIX$:1]: We should have. :thinking:

Comment (jonpas) [reply optionals/norealisticnames/$PBOPREFIX$:1]: Some components have it, some don't, all can and should have it though. Mostly older ones don't have it as those files never get touched.

File: addons/fire/config.cpp [hunk 5]
```diff
@@ -1,5 +1,12 @@
 #include "script_component.hpp"
 
+#if __has_include("\z\ace\addons\nomedical\script_component.hpp")
+#define PATCH_SKIP "No Medical"
+#endif
+
+#ifdef PATCH_SKIP
+ACE_PATCH_NOT_LOADED(ADDON,PATCH_SKIP)
+#else
```
Comment (jonpas) [inline addons/fire/config.cpp:9]:
```suggestion
#if __has_include("\z\ace\addons\nomedical\script_component.hpp")
ACE_PATCH_NOT_LOADED(ADDON,PATCH_SKIP)
#else

```

Why not just this? :thinking:

Comment (Dystopian) [reply addons/fire/config.cpp:9]: This way can define `PATCH_SKIP` with several conditions. E.g.:
```
#if __has_include("\z\ace\addons\nomedical\script_component.hpp")
#define PATCH_SKIP "No Medical"
#endif
#if __has_include("\some\other_addon\config.bin")
#define PATCH_SKIP "Some Other Addon"
#endif
```
and then you use `PATCH_SKIP` just once to disable component with `ACE_PATCH_NOT_LOADED`. In this PR this is mostly a good template for other `__has_include` components. I liked @PabstMirror idea and implemeted it here too.
See also https://github.com/acemod/ACE3/blob/fdbcc62c74f1c51cf3ff36b784789a92fada4702/addons/compat_ws/config.cpp#L3-L9

Also you have `PATCH_SKIP` undefined in suggestion.

Comment (jonpas) [reply addons/fire/config.cpp:9]: Ok your example adds multiple conditions, but you can also do that like this:
```cpp
#if __has_include("\z\ace\addons\nomedical\script_component.hpp")
#if __has_include("\some\other_addon\config.bin")
ACE_PATCH_NOT_LOADED(ADDON,"Some Other Addon")
#else
ACE_PATCH_NOT_LOADED(ADDON,"No Medical")
#else
// default config
#endif
#endif
```
Although that is uglier and I wouldn't want to do it.

Not sure, we just need the opposite of `skipWhenMissingDependencies`!

_Fixed suggestion._

Comment (jokoho48) [reply addons/fire/config.cpp:9]: I am not fully sure about `#if`, but I seem to remember that preprocessor "commands" like that are not able to be nested(at least `#ifdef` is not able to be nested)

Comment (jonpas) [reply addons/fire/config.cpp:9]: Ah, and they also don't support logical operators I believe. Yeah, we need a `skipWhenLoaded`.

Comment (PabstMirror) [reply addons/fire/config.cpp:9]:
```
#if __has_include("\rhsafrf\addons\rhs_main\loadorder\config.bin")
#else
    #define PATCH_SKIP "RHS AFRF"
#endif

#if __has_include("\z\ace\addons\explosives\script_component.hpp")
#else
    #ifdef PATCH_SKIP
    #else
        #define PATCH_SKIP "ACE Explosives"
    #endif
#endif
```
works on game, mikro and hemtt

Comment (rautamiekka) [reply addons/fire/config.cpp:9]: What about [`#ifndef`](https://community.bistudio.com/wiki/PreProcessor_Commands#.23ifndef), is it working everywhere ? Cuz if it is, I don't see why do ifdef-else.

Comment (veteran29) [reply addons/fire/config.cpp:9]: >  but I seem to remember that preprocessor "commands" like that are not able to be nested(at least #ifdef is not able to be nested)

This should be fixed at least for a patch or two.

Comment (LinkIsGrim) [reply addons/fire/config.cpp:9]: #9294 seems to work fine with nesting.

**Review by jonpas — APPROVED** [review]

**Pull request merged**